[mlir][ArmSME] Support vertical layout in load and store ops

c-rhodes opened the pull request:
In SME a ZA tile slice is a one-dimensional set of horizontally or vertically contiguous elements within a ZA tile. Currently the load and store ops only support horizontal tile slices. This patch adds a tile slice layout attribute to the load and store ops to support both horizontal and vertical tile slices.

When lowering from Vector dialect horizontal layout is the default.

c-rhodes committed:
[mlir][ArmSME] Support vertical layout in load and store ops

In SME a ZA tile slice is a one-dimensional set of horizontally or
vertically contiguous elements within a ZA tile. Currently the load and
store ops only support horizontal tile slices. This patch adds a tile
slice layout attribute to the load and store ops to support both
horizontal and vertical tile slices.

When lowering from Vector dialect horizontal layout is the default.

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:
```diff
@@ -204,29 +204,51 @@ struct LoadTileSliceToArmSMELowering
     case 8:
-      rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(loc, allActiveMask, ptr,
-                                                       tileI32, tileSliceI32);
+      if (layout == arm_sme::TileSliceLayout::Horizontal)
+        rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+      else
+        rewriter.create<arm_sme::aarch64_sme_ld1b_vert>(loc, allActiveMask, ptr,
+                                                        tileI32, tileSliceI32);
```
Wondering how to get rid of these `if` statements. You could try this:

```suggestion
      bool isHorizontal = (layout == arm_sme::TileSliceLayout::Horizonta);
      isHorizontal ? 
        rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(
            loc, allActiveMask, ptr, tileI32, tileSliceI32) :
        rewriter.create<arm_sme::aarch64_sme_ld1b_vert>(
            loc, allActiveMask, ptr, tileI32, tileSliceI32);
```

or this:

```cpp
if (layout == arm_sme::TileSliceLayout::Horizontal) {
    switch (tileElementWidth) {
    default:
       llvm_unreachable("unexpected element type!");
    case 8:
    // (...)
    break;
} else {
    switch (tileElementWidth) {
    default:
       llvm_unreachable("unexpected element type!");
    case 8:
    // (...)
    break;
}
}
```

c-rhodes replied:
I've implemented the latter suggestion

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/CMakeLists.txt:
```diff
@@ -4,3 +4,10 @@ add_mlir_doc(ArmSME ArmSME Dialects/ -gen-dialect-doc -dialect=arm_sme)
 mlir_tablegen(ArmSMEConversions.inc -gen-llvmir-conversions)
 add_public_tablegen_target(MLIRArmSMEConversionsIncGen)
+
+mlir_tablegen(ArmSMEEnums.h.inc -gen-enum-decls)
+mlir_tablegen(ArmSMEEnums.cpp.inc -gen-enum-defs)
+mlir_tablegen(ArmSMEAttrDefs.h.inc -gen-attrdef-decls -attrdefs-dialect=arm_sme)
+mlir_tablegen(ArmSMEAttrDefs.cpp.inc -gen-attrdef-defs -attrdefs-dialect=arm_sme)
+add_public_tablegen_target(MLIRArmSMEEnumsIncGen)
+add_dependencies(mlir-headers MLIRArmSMEEnumsIncGen)
```
Why do we need all these changes? (there's quite a lot in here)

c-rhodes replied:
this is for the attribute, see: https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/#cmake-targets

It seems just `add_public_tablegen_target(MLIRArmSMEAttrDefsIncGen)` works tho so replaced bottom two lines with that.

banach-space commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir:
```diff
@@ -0,0 +1,110 @@
+  // to memory and reload to print.
+  vector.store %0, %mem2[%c0] : memref<?xi32>, vector<[4]x[4]xi32>
+
+  // Dump "mem1". The smallest SVL is 128-bits so the tile will be at least
+  // 4x4xi32.
```
I love this test - it does a great job illustrating the concept of horizontal vs vertical loads!

[nit] Worth adding a note that this is the original layout, e.g.:


```suggestion
  // 1. ORIGINAL LAYOUT
  // Dump "mem1". The smallest SVL is 128-bits so the tile will be at least
  // 4x4xi32.
```

c-rhodes replied:
Done

banach-space commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-load-vertical.mlir:
```diff
@@ -0,0 +1,110 @@
+  }
+  func.call @printTileEnd() : () -> ()
+
+  // Dump "mem2". The smallest SVL is 128-bits so the tile will be at least
+  // 4x4xi32.
```
```suggestion
  // 2. VERTICAL LAYOUT
  // Dump "mem2". The smallest SVL is 128-bits so the tile will be at least
  // 4x4xi32.
```

c-rhodes replied:
Done

banach-space commented on mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir:
```diff
@@ -0,0 +1,401 @@
+// RUN: mlir-opt %s -convert-vector-to-llvm="enable-arm-sme" -cse -canonicalize -split-input-file -verify-diagnostics | FileCheck %s
+
+// Test conversion of higher-level ArmSME ops to LLVM intrinsics.
```
By "higher-level" you basically mean the custom ops (as opposed the intrinsics), right? As in, there are 2 categories ATM, correct?

c-rhodes replied:
Not really sure how I can make this any clearer than it is

banach-space replied:
Perhaps just drop "higher-level"? That's the bit that confused me.

c-rhodes replied:
Done

banach-space commented on mlir/test/Dialect/ArmSME/arm-sme-to-llvm.mlir:
```diff
@@ -0,0 +1,401 @@
+// -----
+
+// CHECK-LABEL: @arm_sme_load_tile_slice_hor_i16
+// CHECK: arm_sme.intr.ld1h.horiz
```
I'd be tempted to check the types as well:
```suggestion
// CHECK: arm_sme.intr.ld1h.horiz (%{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}) : (vector<[8]xi1>, !llvm.ptr, i32, i32) -> ()
```

c-rhodes replied:
Done

c-rhodes committed:
reformat switch

c-rhodes committed:
check types in arm-sme-to-llvm.mlir

c-rhodes committed:
add comment to integration test

c-rhodes committed:
replace enumsincgen with attrincgen

banach-space reviewed (approved):
LGTM, thanks!

c-rhodes committed:
Update arm-sme-to-llvm.mlir to drop "higher-level"

dcaballe commented on mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:
```diff
@@ -204,30 +204,59 @@ struct LoadTileSliceToArmSMELowering
+
+    // Create 'arm_sme.intr.ld1*.(horiz|vert)' intrinsic to load ZA tile slice.
+    if (layout == arm_sme::TileSliceLayout::Horizontal) {
+      switch (tileElementWidth) {
+      default:
+        llvm_unreachable("unexpected element type!");
+      case 8:
+        rewriter.create<arm_sme::aarch64_sme_ld1b_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 16:
+        rewriter.create<arm_sme::aarch64_sme_ld1h_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 32:
+        rewriter.create<arm_sme::aarch64_sme_ld1w_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 64:
+        rewriter.create<arm_sme::aarch64_sme_ld1d_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
+      case 128:
+        rewriter.create<arm_sme::aarch64_sme_ld1q_horiz>(
+            loc, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
```
worth refactor this into a template function?

c-rhodes replied:
it's not obvious to me how I could improve this with a template function, could you give us some pointers?

banach-space replied:
This is the only thing that came to my mind:
```cpp
template <int N>
void callLoadIntrinsic(ConversionPatternRewriter &rewriter,
                       arm_sme::StoreTileSliceOp storeTileSliceOp,
                       mlir::vector::SplatOp allActiveMask, Value ptr, Value tileI32,
                       mlir::arith::IndexCastUIOp tileSliceI32) {
}

template <>
void callLoadIntrinsic<8>(ConversionPatternRewriter &rewriter,
                          arm_sme::StoreTileSliceOp storeTileSliceOp,
                          mlir::vector::SplatOp allActiveMask, Value ptr,
                          Value tileI32,
                          mlir::arith::IndexCastUIOp tileSliceI32) {
  rewriter.replaceOpWithNewOp<arm_sme::aarch64_sme_st1b_horiz>(
      storeTileSliceOp, allActiveMask, ptr, tileI32, tileSliceI32);
}
```

But that wouldn't be an improvement IMHO, so I'd go ahead with what you have here already.

dcaballe replied:
You could have a template function with a single switch and then:
`callLoadIntrinsic<arm_sme::aarch64_sme_ld1b_horiz, arm_sme::aarch64_sme_ld1h_horiz...>(...)`
`callLoadIntrinsic<the vertical ones>(...)`

That should reduce the code by 2x but maybe I'm missing something :)

dcaballe commented on mlir/lib/Dialect/ArmSME/Transforms/LegalizeForLLVMExport.cpp:
```diff
@@ -280,29 +309,58 @@ struct StoreTileSliceToArmSMELowering
+      case 128:
+        rewriter.replaceOpWithNewOp<arm_sme::aarch64_sme_st1q_horiz>(
+            storeTileSliceOp, allActiveMask, ptr, tileI32, tileSliceI32);
+        break;
```
same?

dcaballe reviewed (approved):
Awesome! Just a small refactoring comment

c-rhodes committed:
make tile slice layout default

c-rhodes committed:
hor -> horizontal, ver -> vertical

c-rhodes committed:
foo

c-rhodes committed:
foo

c-rhodes commented:
thanks for reviewing all. I've updated this to make tile slice layout optional with a default of horizontal, as suggested by @dcaballe on #66760. Optional custom attributes aren't supported as the first argument in the asm string so I've moved layout to the end of the asm string for both loads and stores for consistency. Quite a lot of churn to support this unfortunately, would appreciate a second pair of eyes on the changes before I land.

c-rhodes commented:
Forgot to mention, also changes the format in the IR from:
`<hor> -> <horizontal>`
`<ver> -> <vertical>`
as suggested on https://github.com/llvm/llvm-project/pull/66760.

c-rhodes committed:
fix linter errors
